BranchMetrics/android-branch-deep-linking-attribution
# feat(threading): Convert Branch.java to Kotlin

## Reference
SDK-1979 -- [Android] Convert Branch.java to Kotlin

## Description
First part of work required to resolve our threading issues.

This PR is a prerequisite for refactoring our current threading and tasks and introducing Kotlin framework-managed concurrency libraries.

There are several motivations for doing so which are covered in an internal doc.

Changes
- Some deprecated methods were causing errors in the automatic conversion, thus were removed as they were deprecated for over a year (sept 2021 and earlier)

- The largest change is the **mechanical** conversion of `Branch.java` to `Branch.kt`. It should be noted that an intermediary commit was made to rename the file (`git mv`) from `Branch.java` -> `Branch.kt` and then converting the code to Kotlin by the tool. Otherwise, the delta is too great for git to correlate the two files. The GitHub PR web interface shows them as separate.
  - There is some reorganizing of static methods into the companion object created so it would appear that some of the code is deleted but is just moved further on down. https://kotlinlang.org/docs/coding-conventions.html#class-layout

- The annotation `@JvmStatic` was added on some fields to remove the requirement of referencing the `Companion` Kotlin object in Java. 
  - For example, after the converting, Java callers would have needed to invoke `Branch.Companion.getAutoInstance(this)` https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/1065/commits/eeee0d86cce7c6f87ea7f54e762481cf29fcba5e#diff-11bfbd9108d99eab0b6fc006502e2a7b29e39cd75bac80d8484cfce7468f19d3L26
  - By adding it, the static methods/fields are generated in the bytecode and the original manner of referencing these static methods are unchanged. `Branch.getAutoInstance(this)`

- Some classes were made `public` because getters on the `Branch` object were `public` and broke the package visibility originally in Java.
  - Class/API visibility will be addressed in a later PR before this feature is finished.
  
- Fixed some unit tests that didn't infer `nullability`. 

## Testing Instructions
1. Checkout the branch.
1a. Ensure git history is preserved in `Branch.kt` with commits prior in Java.
2. Manually verify operations in the testbed app still work as expected in your environment.
3. Integrate by maven local (non-SDK-project module reference)
4. Ensure in your Java test app no contracts have been changed. 
4a. Kotlin will have the getter change from `Branch.getInstance()` -> `Branch.instance`


## Risk Assessment [`MEDIUM`]


Because this was machine translated, it is possible that it made changes confidently incorrect. Effort was made to scan the diff manually. If you contributed, please ensure your changes appear as they should in Kotlin. 

- [✅] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- Correctness & Style

cc @BranchMetrics/saas-sdk-devs for visibility.
